### ISSUE TITLE
Memory issue on a SPA

### ISSUE THREAD

**Issue by mebibou** [AnyChart/AnyChart#41]
I am not sure if it has to do with how we are using anychart in our AngularJS project, but we're experiencing memory leaks on pages where we have some charts displayed. I've created a simple plunkr using angular route and a few charts: https://plnkr.co/edit/tKhqlu?p=preview

If I load the page, then open the stock chart, then go back to the pie chart, and taking heap snapshots on Chrome dev tool along the way, I can see that the memory increases on the stock chart but does not get released on the Pie chart after going back.

![screen shot 2018-04-04 at 1 56 13 pm](https://user-images.githubusercontent.com/305342/38308755-fd5f50ee-380f-11e8-97de-382047246132.png)

From this screenshot, what I did:
1. Open the page, then "Take heap snapshot" > Snapshot 1
2. Click on "Stock chart", then "Take heap snapshot" > Snapshot 2
3. Click on "Pie chart", then "Take heap snapshot" > Snapshot 3

Normally Snapshot 3 should be at 76.6Mb again I think, but somehow it remains more than on Snapshot 2. Is there some function to be called on the chart to cleanly destroy it?

**Comment by Shestac92** [comment]
@mebibou 
In this sample, every time when you switch to stock the code creates a new instance of the stock chart and set to scope:
`$scope.stock = chart;`
Existing stock charts are stored in the heap. This leads to memory leak. 
It's very easy to solve. All you need is to check if the there's an existing stock chart and set to scope a link on the instance of the chart. We have modified your example, now the heap doesn't grow. Please, check [the forked sample](https://plnkr.co/edit/K8xHU0ZqrAkSRuHs6eEw?p=preview).

**Comment by mebibou** [comment]
@Shestac92 so your solution is to keep all the instances of the stocks in memory and never delete the reference? If I display 100 stock charts during a session, I will need to keep 100 charts in memory? no what I want is that when the scope of the chart is destroyed, it should be removed from memory, period. Even if 2 seconds later I load the same one again, it should load it in memory at that point.

**Comment by mebibou** [comment]
I'm guessing this means you also keep references somewhere to the data even if the charts no longer exists? in this case we would really need to be able to do this:
```
$scope.$on('$destroy', function() {
  $scope.chart.destroy();
});
```

**Comment by Shestac92** [comment]
@mebibou 
this workaround is exactly for this sample. If you need to show 100 stock charts you can: create 1 stock chart and change data and appearance, you can dispose it and create a new one. Using the approach from the modified sample allows to save (cache) the instance of the chart as long as you need. Then you can drop the reference. And you'are absolutely right, you can always destroy the `$scope.chart` at any moment too.

**Comment by Shestac92** [comment]
This is Angular behavior. Our library and plugin don't save references on the instance if the chart is no longer in the DOM. And we provided the sample as a workaround to use repeatedly the instance of existing chart and don't create the instance again.

**Comment by mebibou** [comment]
@Shestac92 just so we don't blame Angular for the behavior, I've made a new plunkr that uses pure Vanilla JS: https://plnkr.co/edit/engHIv?p=preview. So again, in this trivial example, how to free up the memory of previously loaded Chart that do not exist in the DOM anymore?

**Comment by mebibou** [comment]
@Shestac92 something I just saw while debugging: when a chart is being added, it adds a `<svg></svg>` element at the end of the `body` tag, but when removing the chart this same element does not get removed from the DOM (in our app I saw about 50 of them there). I though this might be what is causing the leak so I tried removing those elements from the DOM after removing the chart, but to no avail, the memory still does not get released

**Comment by Shestac92** [comment]
@mebibou 
Thank you for this report! This is very interesting and unexpected thing. Probably there's a link between this svg element and memory leak. In any case, we're going to investigate it. I will update you as I get results.

**Comment by Shestac92** [comment]
@aharpervc 
Thank you for these details! Charts provide special method dispose() - which destroys the chart's svg. You can call it like this:
```
this.chart.dispose();
```

**Comment by mebibou** [comment]
@Shestac92 so there is a method to supposably clean up when a chart is removed? I just tried on the plunkr, it does remove the `svg` elements but it does not completely free up memory.

And also I believe it is missing the definitions like you can see on the screenshot
<img width="617" alt="screen shot 2018-04-18 at 8 58 03 am" src="https://user-images.githubusercontent.com/305342/38918983-3607ade2-42e7-11e8-96e0-ffbf70181108.png">

**Comment by Shestac92** [comment]
@mebibou 
Thank you for the report! We will review typedefs and add dispose() method for required classes.
We continue investigating svg elements in DOM issue.

**Comment by Shestac92** [comment]
@mebibou @aharpervc 
We've finished investigating all these issues. 
To remove svg elements from the DOM structure you should call dispose() method. This method will be defined in next release of index.d.ts. Actually, these svg elements don't impact on memory dramatically.
The approach of creating a new instance of chart every time it's showed to a user leads to the memory leak. The better approach is to create a chart when it's required, store the instance if it and set it to scope when the chart is required to show up again.

**Comment by mebibou** [comment]
@Shestac92 the approach to "store the instance" is what leads to memory leak, period. If I have 10 charts on a page, I'm not going to store the 10 charts in memory "just in case" I need to reuse them later, this is nonsense. This is not a better approach, this is an approach a junior JavaScript developer would take to avoid the real problem.
This is not good enough, and for a paid library it is not acceptable.

**Comment by mebibou** [comment]
@Shestac92 I'm not sure if you meant keeping the chart reference and also the data within so all you have to do is simply show the chart as it was before it was removed (that would be even worse since it would only be viable if you only display static charts, otherwise you would keep hundreds of charts in memory), but I tried to play with simply removing data from a chart and adding the data again, and yet the memory still grows: https://plnkr.co/edit/YVnFKd?p=preview

**Comment by blowsie** [comment]
@mebibou I've encountered memory issues around the DOM before. You should consider that when you remove items from the DOM they are not instantly removed from memory. They are only removed when Garbage Collection runs. 

Consider this quote.

> Garbage Collection
You must run garbage collection at the beginning and end of any test run. While a single state change might take 5 MB of heap, that memory should be reclaimed by garbage collection. Normally, it's up to the browser to initiate garbage collection. The browser may choose to delay a full GC for a very long time, and the results might be very misleading. By running GC at the beginning and end of any test run, we can normalize the actual heap memory and nodes in use.

_I am not affiliated with AnyCharts_

**Comment by blowsie** [comment]
The point is, you can manually trigger the Garbage collection from the console, if the memory usage still exists then there is a problem.

**Comment by mebibou** [comment]
@blowsie yes tried that before on all the plunkr I posted here, didnt do anything when clicking on the Trash button in the Performance tab (unless there is another way that does more?)

**Comment by mebibou** [comment]
by didn't do anything I meant the heap snapshot after was the same as before

**Comment by xigolle** [comment]
Is this already resolved picked up ? We are experiencing memory issues in an Angular 2 application with using AnyChart. 
But want to be sure this is already resolved in order to investigate any further.

**Comment by techBeck03** [comment]
I'm curious about this as well.  Any resolution?  Thanks!
